### PR TITLE
Add dyld shared cache parsing

### LINKED ISSUE #268: Plans for macOS Big Sur system libraries in the dyld shared cache?
Hi,

starting with macOS Big Sur, it has become a fair bit harder to obtain symbols from most macOS system libraries, e.g. for profilers or stack traces: The libraries are no longer present as files in the file system, instead they are in a system-wide dyld shared cache file. See [these Google results](https://www.google.com/search?q=big+sur+"dyld+shared+cache") for more information.
So one way to get at the symbols would be to extract the libraries from the cache. Another option, when running inside the process that has the relevant libraries loaded, would be to find the mappings in the process memory and get the symbol table from there.
Do you know if somebody has already started working on rust code to parse the dyld_shared_cache format and write an extractor? Or whether there are any other plans to make getting symbols for macOS system libraries easier in the rust ecosystem?

I imagine this would also be needed to get proper symbols in rust panic stack traces, for example. (Though those seem to be broken at the moment even before Big Sur, see https://github.com/rust-lang/rust/issues/79831 ).

### PR DESCRIPTION
**Pull request by mstange**
Fixes #268.

This PR is based on top of #304. For my purposes, #304 on its own is enough to unblock me, because I can have the dyld cache parsing code outside of object. But it might be worth landing into object anyways.
So this is more of a "please take the parts you like" kind of PR.

### CODE REVIEW

**Comment by philipc** [comment]
I think this is fine to add to this crate. I want to experiment a bit with the API though, so I'll probably edit this and keep you as a coauthor.

**Comment by mstange** [comment]
Sounds good!

File: src/read/macho/dyld_cache.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            "Couldn't read path string (didn't find nul byte within first 256 bytes)",
+        )?;
+        // The path should always be ascii, so from_utf8 should alway succeed.
+        let path = core::str::from_utf8(path).map_err(|_| Error("Path string not valid utf-8"))?;
+        Ok(path)
+    }
+
+    /// The offset in the dyld cache file where this image starts.
+    pub fn offset(&self) -> u64 {
+        self.image_info.address.get(self.endian) - self.first_mapping_address
```
Comment (philipc) [inline src/read/macho/dyld_cache.rs:201]: Is this assuming that all images are in the first mapping?

File: src/read/macho/dyld_cache.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    /// The file system path of this image.
+    pub fn path(&self) -> Result<&'data str> {
+        // The longest path I've seen is 164 bytes long. In theory paths could be longer than 256.
+        const MAX_PATH_LEN: u64 = 256;
+
+        let path_offset = self.image_info.path_file_offset.get(self.endian) as u64;
+        let slice_containing_path = self
+            .data
+            .read_bytes_at(path_offset, MAX_PATH_LEN)
+            .read_error("Couldn't read path")?;
```
Comment (philipc) [inline src/read/macho/dyld_cache.rs:190]: What do you think about adding a `ReadRef::read_string_at` or something like that?

File: src/read/macho/dyld_cache.rs (hunk 1 above)
Comment (mstange) [reply src/read/macho/dyld_cache.rs:201]: I'm not sure - this is based on the code in [dsc_iterator.cpp](https://opensource.apple.com/source/dyld/dyld-851.27/dyld3/shared-cache/dsc_iterator.cpp.auto.html), specifically the function `forEachDylibInCache` which calls this offset `firstRegionAddress`. It applies the same offset to all images.

File: src/read/macho/dyld_cache.rs (hunk 2 above)
Comment (mstange) [reply src/read/macho/dyld_cache.rs:190]: Sounds good to me.

File: src/read/macho/dyld_cache.rs (hunk 1 above)
Comment (mstange) [reply src/read/macho/dyld_cache.rs:201]: I read your implementation and I agree it looks more correct. To make sure it doesn't break anything, I looked at a few dyld shared cache files. The "correct" code is equivalent to the simple code if there is only one mapping, or if all the images are in the first mapping, or if all the mappings containing images have the same value for `mapping.address - mapping.offset`.
It seems like the second option is the case: In the cache files I've checked, all images are in the first mapping. The x86_64 and x86_64h caches on 10.14.6 and on 11.3.1 each have three mappings, and the arm64e cache on 11.3.1 has seven mappings. The mappings have different values for `mapping.address - mapping.offset`, but that's fine since all images are in the first mapping.
So, in summary, what you have looks great and shouldn't break anything.
